rust/connlib/tunnel/src/client.rs

```diff
@@ -109,6 +109,30 @@ where
         Ok(())
     }
 
+    /// Sets the dns server.
+    pub fn set_upstream_dns(&self, upstream_dns: &[DnsServer]) {
+        let upstream_dns: HashSet<_> = upstream_dns
+            .iter()
+            .map(|dns| {
+                let DnsServer::IpPort(dns) = dns;
+                dns.address
+            })
+            .collect();
+
+        // TODO: assuming single dns
+        let Some(upstream_dns_srv) = upstream_dns.iter().next().cloned() else {
+            return;
+        };
+
+        self.role_state.lock().upstream_dns = upstream_dns;
+
+        self.role_state
+            .lock()
```

Comment:
Why is `role_state` locked twice back-to-back here?

Reply:
I rather keep locks around for as little as possible, to reduce contention. There's no need to keep any syncing between the two times `role_state` is locked in this function.

Reply:
They are both DNS-related but okay

Reply:
But, if `upstream_dns` is modified and another thread modifies `peers_by_ip` before this can there will be no harm since the created peer will use the new `upstream_dns`